### ISSUE TITLE
How to correctly decorate a stateful criterion?

### ISSUE THREAD

**Issue by hsheil** [Element-Research/rnn#120]
Hi

_[For now I'm using BCECriterion as my classification problem is binary and this is stateless so works with SequencerCriterion, but I do want to scale the code out to do multi-class, so would like to go back to ClassNLLCriterion eventually..]_

In extending the code written for #92 to a real-world data set I've hit a problem with using nn.ClassNLLCriterion() as follows. I can see why the errors are happening and understand the logic, the issue is that it seems as if SequencerCriterion is not compatible with the criterion (ClassNLLCriterion) I want to use as I'm trying to solve a classification problem. The issue is described as follows:

```lua
--gives the error 'SequencerCriterion only decorates stateless criterions : nn.ClassNLLCriterion'
--when we call seqC:backward(out, batchTargets) later on in the code
local crit = nn.ClassNLLCriterion()
local seqC = nn.SequencerCriterion(crit)
```

Then looking at #93 I tried:
```lua
--gives 'SequencerCriterion shouldn't decorate a ModuleCriterion. Instead, try the other way around : 
--ModuleCriterion decorates a SequencerCriterion. Its modules can also be similarly decorated with a 
--Sequencer.' - thrown in the constructor..
local seqC = nn.SequencerCriterion(nn.ModuleCriterion(nn.ClassNLLCriterion(), nn.LogSoftMax()))
```

Then I looked for "ModuleCriterion" in the repo to find a working example and I see code similar to my original code which I think *should* work, e.g. https://github.com/Element-Research/rnn/blob/442276f1e185867f22744386096eb6320a1362f6/examples/recurrent-language-model.lua#L131-#L135

All insights appreciated into what I'm missing?!

**Comment by nicholas-leonard** [comment]
@hsheil I cannot reproduce your error with `nn.SequencerCriterion(nn.ClassNLLCriterion())`. I updated my nn package to see if your error wasn't caused by changed to nn. This criterion is explicitly unit tested to be stateless (see test/test.lua).

For your second issue, do as in the example you pointed to, i.e. the ModuleCriterion wraps the SequencerCriterion; you are doing the opposite :).

**Comment by hsheil** [comment]
@nicholas-leonard I have to take it back - I luarocks installed everything again and the original code seems to be working.

Almost have code ready for part two, just a weird error I have to debug ("Sequencer.lua:73: gradOutput should have as many elements as input"). In this case gradOutput from rnn:forward(batchInputs) is always 0. Here's a code snippet:

```lua
   -- For this data set, a single element of the batch is 1..* session events
   -- so 'batchSize = 5' means give me all of the events for the next 5 sessions in the yoochoose dataset
   ds.setBatchSize(batchSize)
   --the ds is a cursor into our training set, giving back a batch each time it is called
   --next() operates like / is inspired by a JDBC ResultSet e.g. https://docs.oracle.com/javase/8/docs/api/java/sql/ResultSet.html#next--
   while ds.next() do
      local start =torch.tic()
      local batchInputs = {}
      local batchTargets = {}
      batchInputs, batchTargets = ds.getData()
      local out = rnn:forward(batchInputs)
      print('out size is ' ..#out)
      err = err + seqC:forward(out, batchTargets)
      print('err is ' .. err)
      gradOut = seqC:backward(out, batchTargets)
      print('gradOut size is ' .. #gradOut .. ' and batchInputs size is ' .. #batchInputs)
      rnn:backward(batchInputs, gradOut)
      rnn:updateParameters(opt.lr)
      rnn:zeroGradParameters()
      local currT = torch.toc(start)
      print('loss', err .. ' in ', currT .. ' s')
   end
```

And the output:

Moving data to CUDA for training	
Training phase starting	
out size is 0	
err is 0	
gradOut size is 0 and batchInputs size is 2

**Comment by nicholas-leonard** [comment]
@hsheil can you print batchInputs, batchTargets?

**Comment by hsheil** [comment]
Hi @nicholas-leonard I figured it out - I wasn't batching the individual events correctly so was creating a tensor of 187x1 for each table entry (187 is soon to grow to 53,272 when all of the attributes are in..) when it should have been of dimension 187xn where n is the variable number of events for that session.

**Comment by hsheil** [comment]
Hi @nicholas-leonard I got this error again and given that ClassNLLCriterion should work, I played around with the data I was feeding to the RNN. Turns out that input with a length of 1 and a batchSize of >= 3 triggers this if statement to fire with a false positive of a stateful criterion (I think..). I found this when debugging something else (why increasing batchSize causes loss to explode).

```lua
if #inputTable >= 3 and not self.isStateless then
      -- make sure the criterion is stateless
      local gradInput
      for i = 1,3 do
         self.criterion:forward(inputTable[i], targetTable[i])
         gradInput = self.criterion:backward(inputTable[i], targetTable[i])
         nn.utils.recursiveAdd(gradInput -1, self.gradInput[i])
         if math.abs(nn.rnn.recursiveSum(gradInput)) < 0.0001 then
            error("SequencerCriterion only decorates stateless criterions : "..tostring(self.criterion))
         end
      end
      self.isStateless = true -- test should only be run once
   end
```

**Comment by nicholas-leonard** [comment]
@hsheil Can you create a snippet of code that fails for your use case so that I can reproduce on my end? Also, #inputTable is not the batch size, but the sequence length.

**Comment by hsheil** [comment]
@nicholas-leonard Will do. See my other reply on the #inputTable :)

**Comment by hsheil** [comment]
@nicholas-leonard I think the data creation step - creating the inputs {} and targets {} that RNN needs is a key part to understanding RNN / using it effectively. An animation / graphic could help, possibly inspired by something like this one: http://www.acnc.com/raidedu/0. We could have a crack at it here if you like, maybe using d3 / c3 (http://d3js.org/ / http://c3js.org/)?

**Comment by nicholas-leonard** [comment]
@hsheil You are right! Will look into it this afternoon.

**Comment by trevorfiez** [comment]
I just ran into the sequencer, classnll error on my own and it was because the number of outputs into the criterion was not equal to the number of classes in my output. Could you post the code for your neural network?

Thanks!

**Comment by jrich9999** [comment]
[image: Inline image 1]
http://peterroelants.github.io/posts/rnn_implementation_part02/

**Comment by nicholas-leonard** [comment]
@jrich9999 ?

**Comment by jrich9999** [comment]
@hsheil <https://github.com/hsheil> @nicholas-leonard
<https://github.com/nicholas-leonard> I saw Humphrey's earlier message - "I
think the data creation step - creating the inputs {} and targets {} that
RNN needs is a key part to understanding RNN / using it effectively. An
animation / graphic could help, etc.."  I know that a graphic on the data
creation or taking live data and getting in the right form to feed the
batch processing would be a big help to a new guy like me.  I know I
stumbled on that and still have some real gray areas of understanding.  I
guess it is only natural when you use a framework at an abstracted level.
Unless you dig deep into how the framework works, you need some good
guidance on how to set thing up right to feed it.  I had saw petter's (
http://peterroelants.github.io/posts/rnn_implementation_part02/) and
figured I would reference it, though it is for python RNN, it seamed
somewhat relevant.  But I'm new, sorry if I am off base...  I do love Torch
and the Element-Research's additions and you guys for your time and
guidance...

**Comment by nicholas-leonard** [comment]
So I added some documentation with images in #131. The new doc is visible here : https://github.com/Element-Research/rnn#inputoutput-format . Hope that works?
